### PR TITLE
GH action updates

### LINKED ISSUE #492: Add to documentation instructions for bug fix release
Add to documentation instructions for bug fix release, with `git cherry-pick`. This can get inspired by these instructions for qutip, https://github.com/qutip/qutip-doc/blob/master/release_distribution.rst#apply-bug-fix-to-latest-release.

### PR DESCRIPTION
**Pull request by crazy4pi314**
Description
-----------
Closes #492, #493, #531, and #391.

Adds:
- New greeting action for when a new contributor to Mitiq files and issue or PR to let them know what will happen next and thank them for their contribution. (`greeting.yml`)
- Nightly pushes of code on master to TestPyPI. (`publish-testpypi.yml`)
- Automated pushing of package to PyPI when a release on GH is made. (`publish-pypi.yml`)
- Pre-release build tests/format checking on PRs made from master to release. (`pre-release.yml`)
- Automatic draft releases created when sematic version tag is created. (`gh-release.yml`)

Modifies:
- Existing CI workflows work basically the same (tests that fire when a PR is made to Master, and when stuff is pushed to Master), but have been renamed to better describe what they do, and reduced test time for the format checking ( there was duplicate environments setup for when just linting tests were being run.) (`test.yml` and`test-master.yml`)

My credentials have been added as repo secrets for the PyPI pushing FYI, we may want to create a bot account for such.

Checklist
-----------
Please make sure you have finished the following tasks before requesting a review of the pull request (PR). For more information, please refer to the [Contributor's Guide](../blob/master/CONTRIBUTING.md):

- [x] Contributions to mitiq should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/). You can enforce it easily with [`black`](https://black.readthedocs.io/en/stable/index.html) style and with [`flake8`](http://flake8.pycqa.org) conventions.
- [x] Please add tests to cover your changes, if applicable, and make sure that all new and existing tests pass locally.
- [x] If the behavior of the code has changed or new feature has been added, please also [update the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md).
- [x] Functions and classes have useful [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings and [type hints](https://www.python.org/dev/peps/pep-0484/) in the signature of the objects.
- [x] (Bug fix) The associated issue is referenced above using [auto-close keywords](https://help.github.com/en/articles/closing-issues-using-keywords). If the PR fixes an issue, use the keyword fix/fixes/fixed followed by the issue id. For example, if the PR fixes issue 1184, type "Fixes #1184" (without quotes).
- [x] The [changelog](https://github.com/unitaryfund/mitiq/blob/master/CHANGELOG.md) is updated, including author and PR number (@username, gh-xxx).

### CODE REVIEW

**Review by andreamari — APPROVED** [review]
LGTM, even though I have no idea if this is going to work in practice :smiley: !

P.s. see comment about `testpypi`.

File: .github/workflows/publish-testpypi.yml
Comment (andreamari) [inline .github/workflows/publish-testpypi.yml:33]:
```suggestion
        twine upload --repository testpypi dist/*
```
I guess this should be like this.

**Comment by crazy4pi314** [comment]
@nathanshammah I have added more detail here such that I think it could be ok to review this as one batch, most of the changes are minor and non-interacting. I have tested the test workflows  (which are basically identical to b4).

File: .github/workflows/greeting.yml
Comment (nathanshammah) [inline .github/workflows/greeting.yml:1]: At the moment this does not seem to add too much to the experience in my opinion. I understand that the checks are done behind the scenes at https://github.com/actions/first-interaction. I'm neutral about it. Just wonder if we could add more content if we need to add this. Maybe a link to Discord's mitiq channel.

File: .github/workflows/pre-release.yml
Comment (nathanshammah) [inline .github/workflows/pre-release.yml:72]: Is this finished?

File: .github/workflows/publish-pypi.yml
Comment (nathanshammah) [inline .github/workflows/publish-pypi.yml:32]: I understand this is connected with @crazy4pi314 account, with secrets added to the repo. I think it would be better to create an additional maintainer for this See e.g., bot. https://pypi.org/project/cirq/, https://pypi.org/project/tf-nightly/.

File: .github/workflows/publish-testpypi.yml
Comment (nathanshammah) [inline .github/workflows/publish-testpypi.yml:30]: Same as for PyPI: It may be nice to create a new user for this.

File: .github/workflows/publish-testpypi.yml
Comment (nathanshammah) [inline .github/workflows/publish-testpypi.yml:50]: We could make an issue about this. But for the moment, assuming that if the previous step does not give an error the release has been successful, this could be tested against mitiq without "==x.y.z" for the moment. 

By the way, I think the way to go would be to do something like in qutip, where the the dev version is written (also more useful than the simple `0.x.0dev` message we have now from mitiq.about()) https://github.com/qutip/qutip/blob/master/setup.py#L115

File: .github/workflows/test-master.yml
Comment (nathanshammah) [inline .github/workflows/test-master.yml:29]:
```suggestion
        - name: Run the Mitiq unit tests via pytest
```

File: .github/workflows/test.yml
Comment (nathanshammah) [inline .github/workflows/test.yml:24]:
```suggestion
      # - name: Check types with mypy
```

File: docs/source/release.rst
Comment (nathanshammah) [inline docs/source/release.rst:40]:
```suggestion
code for the release passes them 
```

File: docs/source/release.rst
Comment (nathanshammah) [inline docs/source/release.rst:41]:
```suggestion
 If you have to update the changelog and the version number, do so as a 
```

File: docs/source/release.rst
Comment (nathanshammah) [inline docs/source/release.rst:71]:
```suggestion
Tag the new commit to ``release`` (using ``git tag``) with a tag that matches the
```

Comment (nathanshammah) [reply docs/source/release.rst:71]: I understand ``release`` will be the name of the branch for all of them. If not, maybe this could be clarified.

File: docs/source/release.rst
Comment (nathanshammah) [inline docs/source/release.rst:86]: Does one need to publish the draft release?

File: .github/workflows/publish-testpypi.yml
Comment (nathanshammah) [inline .github/workflows/publish-testpypi.yml:9]: Can you comment what this bit is for?

File: docs/source/release.rst
Comment (nathanshammah) [reply docs/source/release.rst:71]: Does the tagging automatically prompt the PR merging, or one needs to merge it? This could be specified.

**Review by nathanshammah — CHANGES_REQUESTED** [review]
Excited about this, thanks @crazy4pi314! Left a couple of comments.

File: .github/workflows/publish-pypi.yml
Comment (crazy4pi314) [reply .github/workflows/publish-pypi.yml:32]: I have updated the repo secrets to use a bot account (UnitaryFundBot) for both PyPI and TestPyPI workflows, does this work?

File: .github/workflows/publish-testpypi.yml
Comment (crazy4pi314) [reply .github/workflows/publish-testpypi.yml:9]: This is the trigger that will run this job once a day at midnight. It uses standard linux cron time description, I used the [GH docs](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#onschedule) and [this tool](https://crontab.guru/#0_0_*_*_*) to verify the time that I wanted

File: .github/workflows/publish-testpypi.yml
Comment (crazy4pi314) [reply .github/workflows/publish-testpypi.yml:30]: See above, made a bot.

File: docs/source/release.rst
Comment (crazy4pi314) [reply docs/source/release.rst]: I have added more clarification about both of these.

File: docs/source/release.rst
Comment (crazy4pi314) [reply docs/source/release.rst:95]: Yup, added a note.

**Review by nathanshammah — APPROVED** [review]

**Pull request merged**
